src/echoclient/EchoClient.php

```diff
@@ -62,51 +66,60 @@ function __construct()
-                $props);
+        try {
+            $event = new \echoclient\EchoEvent($class, $source, $props, $userId, $timestamp);
+            return $this->sendBatchEvents([$event]);
+        } catch (\Exception $e) {
             return false;
         }
+    }
 
-        $eventUrl = $baseUrl . '/events';
-        $eventJson = $this->getEventJson($class, $source, $props, $userId);
+    /**
+     * Adds multiple events to echo
+     *
+     * @param EchoEvent[] $events An array of EchoEvent objects
+     * @return bool True if successful
+     * @throws InvalidArgumentException If the event data is malformed in any way
+     * @throws EchoHttpException If the server responded with an error
+     * @throws EchoCouldNotSendException If we were unable to send data to the Echo server
+     */
+    public function sendBatchEvents(array $events)
+    {
+        if (empty($events)) {
+            return true;
+        }
 
-        try {
-            $client = $this->getHttpClient();
-            $request = $client->post($eventUrl, $this->getHeaders(), $eventJson, array('connect_timeout' => 2));
-            $response = $request->send();
+        if (count($events) > self::ECHO_MAX_BATCH_EVENTS) {
+            $this->getLogger()->warning('Batch contains more than ' . self::ECHO_MAX_BATCH_EVENTS . ' events');
+            throw new \InvalidArgumentException("Batch of events exceeds the maximum allowed size");
+        }
 
-            if ($response->isSuccessful()) {
-                $this->getLogger()->debug('Success sending event to echo - ' . $class, $props);
-                return true;
-            } else {
-                $this->getLogger()->warning('Failed sending event to echo - ' . $class, array(
-                    'responseCode' => $response->getStatusCode(),
-                    'responseBody' => $response->getBody(true),
-                    'requestProperties' => $props
-                ));
-                return false;
+        foreach ($events as $event) {
+            if (!$event instanceof \echoclient\EchoEvent) {
+                throw new \InvalidArgumentException("Batch must only contain EchoEvent objects");
             }
-        } catch (\Exception $e) {
-            // For any exception issue, just log the issue and fail silently.  E.g. failure to connect to echo server, or whatever.
-            $this->getLogger()->warning('Failed sending event to echo - ' . $class,
-                array('exception' => get_class($e), 'message' => $e->getMessage(), 'requestProperties' => $props));
-            return false;
         }
+
+        $eventsJson = json_encode($events, true);
+
+        // strlen returns no. bytes in a string.
+        $sizeOfBatch = $this->getStringSizeInBytes($eventsJson);
+        if ($sizeOfBatch > self::ECHO_MAX_BATCH_SIZE_IN_BYTES) {
+            throw new \InvalidArgumentException("Batch must be less than 1mb in size");
```

Comment:
This code seems a little hard to develop against. The user of the method would need to create logic to attempt to send a batch of events, catch the failure if one occurs, reduce the batch size and try again. 

Would it be a good idea to provide a batching method so that the user doesn't need to worry about maximum sizes of length and size?

Reply:
It might be good to that, but I think that's beyond the current requirement.

Reply:
This code will need to exist either here or where you plan on using this code. If it isn't here we will get x times implementations.

Reply:
Can you show how you plan on using it?

Reply:
I was working under the assumption that the applications that needed to batch events would deal with that how they see fit.  My difficulty here is that it's not been made clear that making the client store and batch internally is a requirement. 

I think its a sensible idea, but it's not something the node client handles either. I was aiming for parity and this implementation achieves that.  I think we could raise a subsequent ticket to improve this behaviour.

Reply:
Although this seems a good idea on the face of it, it's not obvious to the client what the limits are.  If I try to send 1m events, I'll be waiting a while, and perhaps it'd just be better to say "maybe not?"